### PR TITLE
Fixes #20126 - Enable selective reboot and autoprovision

### PR DESCRIPTION
**Pull request by rabajaj0509**

### CODE REVIEW

File: app/helpers/discovered_hosts_helper.rb [hunk 1]
```diff
@@ -31,9 +31,11 @@ def discovered_hosts_title_actions(host)
   end
 
   def multiple_discovered_hosts_actions_select
-    actions = [[_('Delete hosts'), multiple_destroy_discovered_hosts_path, hash_for_multiple_destroy_discovered_hosts_path]]
+    actions = [[_('Delete Hosts'), multiple_destroy_discovered_hosts_path, hash_for_multiple_destroy_discovered_hosts_path]]
```
Comment (lzap) [inline app/helpers/discovered_hosts_helper.rb]: Please remove "hosts/Hosts" completely.

File: app/helpers/discovered_hosts_helper.rb [hunk 2]
```diff
@@ -31,9 +31,11 @@ def discovered_hosts_title_actions(host)
...
     actions <<  [_('Assign Organization'), select_multiple_organization_discovered_hosts_path,  hash_for_select_multiple_organization_discovered_hosts_path] if SETTINGS[:organizations_enabled]
     actions <<  [_('Assign Location'), select_multiple_location_discovered_hosts_path,  hash_for_select_multiple_location_discovered_hosts_path] if SETTINGS[:locations_enabled]
+    actions <<  [_('Reboot Hosts'), multiple_reboot_discovered_hosts_path, hash_for_multiple_reboot_discovered_hosts_path]
+    actions <<  [_('Auto Provision Hosts'), multiple_auto_provision_discovered_hosts_path, hash_for_multiple_auto_provision_discovered_hosts_path]
```
Comment (lzap) [inline app/helpers/discovered_hosts_helper.rb]: Remove "Hosts" from both actions, let's just keep verb everywhere.

Also re-order the menu in this order please:

* Auto Provision
* Reboot
* Assign Org
* Assign Loc
* Delete

File: lib/foreman_discovery/engine.rb [hunk 3]
```diff
@@ -65,9 +65,9 @@ class Engine < ::Rails::Engine
             :discovered_hosts          => [:update_multiple_location,
                                            :select_multiple_organization,
                                            :update_multiple_organization,
-                                           :select_multiple_location, :refresh_facts, :reboot, :reboot_all],
+                                           :select_multiple_location, :refresh_facts, :reboot, :multiple_reboot],
             :hosts                     => [:process_hostgroup],
-            :"api/v2/discovered_hosts" => [:refresh_facts, :reboot, :reboot_all]
+            :"api/v2/discovered_hosts" => [:refresh_facts, :reboot, :multiple_reboot]
```
Comment (lzap) [inline lib/foreman_discovery/engine.rb]: In API context there is no multiple support, our API can only reboot once host and reboot all. Please keep V2 API unchanged in this regard.

We are actually missing reboot_all CLI command from our CLI, so only API for reboot_all and auto_prov_all can be used, not CLI. If you want go ahead and create these options (`--all`) in https://github.com/theforeman/hammer-cli-foreman-discovery

**Review by lzap — CHANGES_REQUESTED** [review]
Some small nitpick and one comment about the API.

**Comment by sean797** [comment]
[test]

**Comment by lzap** [comment]
Tests finally fixed [test].

Are we missing tests here?

**Comment by lzap** [comment]
Tests are failing because there are not tests fixed and added here. Code looks good, go ahead.

**Comment by lzap** [comment]
There are failures which seems related, do you see them locally?

**Comment by rabajaj0509** [comment]
@lzap yeah, i can see the relevant tests failing, will fix them :)

**Comment by lzap** [comment]
Let's re [test]

**Comment by rabajaj0509** [comment]
@lzap the tests for this pr will only pass after the legacy api is cleaned from discovery : https://github.com/theforeman/foreman_discovery/pull/374

**Comment by lzap** [comment]
We can hopefully [test] now.

**Comment by lzap** [comment]
Please rebase, getting errors:

```
 1) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_all_success:
ActionController::UrlGenerationError: No route matches {:action=>"auto_provision_all", :controller=>"api/v2/discovered_hosts"}
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:185:in `test_auto_provision_all_success'

  2) Error:
Api::V2::DiscoveredHostsControllerTest#test_auto_provision_all_no_rule_success:
ActionController::UrlGenerationError: No route matches {:action=>"auto_provision_all", :controller=>"api/v2/discovered_hosts"}
    /home/lzap/work/foreman_discovery/test/functional/api/v2/discovered_hosts_controller_test.rb:196:in `test_auto_provision_all_no_rule_success'

  3) Failure:
DiscoveredHostsControllerTest#test_multiple_auto_provision_no_rule_success [/home/lzap/work/foreman_discovery/test/functional/discovered_hosts_controller_test.rb:251]:
Expected "No hosts selected" to be nil.

  4) Failure:
DiscoveredHostsControllerTest#test_multiple_reboot_failure [/home/lzap/work/foreman_discovery/test/functional/discovered_hosts_controller_test.rb:269]:
not all expectations were satisfied
unsatisfied expectations:
- expected exactly once, not yet invoked: #<AnyInstance:ForemanDiscovery::NodeAPI::PowerService>.reboot(any_parameters)
satisfied expectations:
- allowed any number of times, not yet invoked: Webpack::Rails::Manifest.asset_paths(any_parameters)
- allowed any number of times, not yet invoked: #<AnyInstance:ForemanDiscovery::NodeAPI::PowerService>.reboot(any_parameters)


  5) Failure:
DiscoveredHostsControllerTest#test_multiple_reboot_error [/home/lzap/work/foreman_discovery/test/functional/discovered_hosts_controller_test.rb:279]:
not all expectations were satisfied
unsatisfied expectations:
- expected exactly once, not yet invoked: #<AnyInstance:ForemanDiscovery::NodeAPI::PowerService>.reboot(any_parameters)
satisfied expectations:
- allowed any number of times, not yet invoked: Webpack::Rails::Manifest.asset_paths(any_parameters)
- allowed any number of times, not yet invoked: #<AnyInstance:ForemanDiscovery::NodeAPI::PowerService>.reboot(any_parameters)


  6) Failure:
DiscoveredHostsControllerTest#test_multiple_auto_provision_success [/home/lzap/work/foreman_discovery/test/functional/discovered_hosts_controller_test.rb:233]:
Expected "No hosts selected" to be nil.
```

The PR you refer to was merged, we can go ahead and finish this, thanks a lot!

**Comment by rabajaj0509** [comment]
Facing this issue, not able to get over this: 

```
bin/rails test _discovery/test/functional/discovered_hosts_controller_test.rb:226

.F

Failure:
DiscoveredHostsControllerTest#test_multiple_reboot_failure [/home/vagrant/foreman_discovery/test/functional/discovered_hosts_controller_test.rb:277]:
Expected response to be a <3XX: redirect>, but was a <403: Forbidden>
```

**Comment by ik5** [comment]
@rahulbajaj0509 , I'ved changed `session: set_session_user_default_manager` to `session: set_session_user(User.current)` and it works.

Apparently `set_session_user_default_manager` does not set it as a manager.

**Comment by rabajaj0509** [comment]
@ik5  thanks a lot! I was stuck with this for a while and your comment was very helpful :)
@lzap we should merge this before we go forward for merging https://github.com/theforeman/foreman_discovery/pull/380, so that we can make appropriate changes there.

**Review by lzap — APPROVED** [review]
Nice, code looks good.

**Comment by lzap** [comment]
Tests are failing, you removed "Reboot All" button but it's being tested in integration tests.

**Comment by rabajaj0509** [comment]
@lzap  looks good now?

**Review by lzap — APPROVED** [review]
Thanks green locally.

File: lib/foreman_discovery/engine.rb [hunk 4]
```diff
@@ -66,7 +66,7 @@ class Engine < ::Rails::Engine
             :discovered_hosts          => [:update_multiple_location,
                                            :select_multiple_organization,
                                            :update_multiple_organization,
-                                           :select_multiple_location, :refresh_facts, :reboot, :reboot_all],
+                                           :select_multiple_location, :refresh_facts, :reboot, :multiple_reboot],
             :hosts                     => [:process_hostgroup],
             :"api/v2/discovered_hosts" => [:refresh_facts, :reboot, :reboot_all]
           }, :resource_type => 'Host'
```
Comment (lzap) [inline lib/foreman_discovery/engine.rb:72]: This is missing some actions, I don't understand why I dont see failures locally but on jenkins:

AccessPermissionsTest.test_0739_route discovered_hosts/submit_multiple_reboot should have a permission that grants access
AccessPermissionsTest.test_0740_route discovered_hosts/submit_multiple_auto_provision should have a permission that grants access

http://ci.theforeman.org/job/test_plugin_matrix/5117/database=postgresql,ruby=2.4,slave=fast/

Comment (lzap) [reply lib/foreman_discovery/engine.rb:72]: Oh it does fail locally, cool. Let me fix this.

Comment (lzap) [reply lib/foreman_discovery/engine.rb:72]: Merged as 84d8e45

**Pull request merged**